### ISSUE TITLE
20221018 auth_role_moderation must be robust against bad data

### ISSUE THREAD

**Issue by robnagler** [radiasoft/sirepo#4844]
```txt
05:00:07 sirepo/uri_router.py:110:call_api api={'base_uri': 'save-moderation-reason', 'cls': <class 'sirepo.auth_role_moderation.API'>, 'decl_ur
i': '/save-moderation-reason', 'func': <function Spec.__call__.<locals>._wrapper at 0x7f023d00e158>, 'func_name': 'api_saveModerationReason', 'name': 'saveModerationReason', 'params': []} exception='PKDict' object has no attribute 'reason' stack=Traceback (most recent call last):
05:00:07 sirepo/uri_router.py", line 102, in call_api
```

If no reason is given, then a uniqueness constraint is raised, which doesn't make sense:

```txt
05:00:13 sirepo/auth_role_moderation.py:140:api_saveModerationReason Error=(sqlite3.IntegrityError) UNIQUE constraint failed: user_role_invite_t
.uid, user_role_invite_t.role
05:00:13 [SQL: INSERT INTO user_role_invite_t (uid, role, status, token, moderator_uid) VALUES (?, ?, ?, ?, ?)]
05:00:13 [parameters: ('<u1>, 'sim_type_jupyterhublogin', <ModerationStatus.PENDING: 'pending'>, '<token>', None)]
```

First step is to create a test for this.

**Comment by robnagler** [comment]
```txt
14:33:47 sirepo/uri_router.py:110:call_api api={'base_uri': 'save-moderation-reason', 'cls': <class 'sirepo.auth_role_moderation.API'>, 'decl_uri': '/save-moderation-reason', 'func': <function Spec.__call__.<locals>._wrapper at 0x7fe43ff21510>, 'func_name': 'api_saveModerationReason', 'name': 'saveModerationReason', 'params': []} exception='PKDict' object has no attribute 'reason' stack=Traceback (most recent call last):
14:33:47 sirepo/uri_router.py", line 102, in call_api
14:33:47 r = flask.make_response(getattr(f.cls(), f.func_name)(**kwargs))
14:33:47 sirepo/api.py", line 82, in _wrapper
14:33:47 return self.func(*args, **kwargs)
14:33:47 sirepo/auth_role_moderation.py", line 152, in api_saveModerationReason
14:33:47 reason=d.reason,
14:33:47 pykern/pkcollections.py", line 69, in __getattr__
14:33:47 f"'{type(self).__name__}' object has no attribute '{name}'"
14:33:47 AttributeError: 'PKDict' object has no attribute 'reason'
14:33:47 Exception was printed at:
14:33:47 AttributeError: 'PKDict' object has no attribute 'reason'
14:33:47 flask/app.py", line 2091, in __call__
14:33:47 return self.wsgi_app(environ, start_response)
14:33:47 flask/app.py", line 2073, in wsgi_app
14:33:47 response = self.full_dispatch_request()
14:33:47 flask/app.py", line 1516, in full_dispatch_request
14:33:47 rv = self.dispatch_request()
14:33:47 flask/app.py", line 1502, in dispatch_request
14:33:47 return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
14:33:47 sirepo/uri_router.py", line 333, in _dispatch
14:33:47 return call_api(route, kwargs)
14:33:47 sirepo/http_reply.py:289:_gen_exception_error unsupported exception=<class 'AttributeError'> msg='PKDict' object has no attribute 'reason'
```

**Comment by robnagler** [comment]
Exceptions like this should not be printed:

```txt
ip0 - - "GET /static/fonts/glyphicons-halflings-regular.woff2 HTTP/1.1" 304 0 "http://v.radia.run:8080/static/css/ext/bootstrap-3.3.7.min.css?20221024.164553" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/ip1 Safari/537.36" "-"
../sirepo/job_api.py:239:request api=api_beginSession runDir=None
sirepo/job_supervisor.py:233:receive ServerReq(api_beginSession, None)
sirepo/job_supervisor.py:1066:__init__ _Op(begin_session, vap6) runDir=None
../sirepo/auth_role_moderation.py:143:api_saveModerationReason Error=(sqlite3.IntegrityError) UNIQUE constraint failed: user_role_invite_t.uid, user_role_invite_t.role
[SQL: INSERT INTO user_role_invite_t (uid, role, status, token, moderator_uid) VALUES (?, ?, ?, ?, ?)]
[parameters: ('cdnqkXeR', 'sim_type_jupyterhublogin', <ModerationStatus.PENDING: 'pending'>, 'ELzJHoSXZQdfft5FPLyjxEisDoUlBBrf', None)]
(Background on this error at: http://sqlalche.me/e/14/gkpj) saving UserRoleInvite for uid=uid2 role=sim_type_jupyterhublogin stack=Traceback (most recent call last):
  File "/home/vagrant/.pyenv/versions/3.7.2/envs/py3/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1706, in _execute_context
    cursor, statement, parameters, context
  File "/home/vagrant/.pyenv/versions/3.7.2/envs/py3/lib/python3.7/site-packages/sqlalchemy/engine/default.py", line 681, in do_execute
    cursor.execute(statement, parameters)
sqlite3.IntegrityError: UNIQUE constraint failed: user_role_invite_t.uid, user_role_invite_t.role

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/vagrant/src/radiasoft/sirepo/sirepo/auth_role_moderation.py", line 135, in api_saveModerationReason
    token=sirepo.util.random_base62(32),
  File "/home/vagrant/src/radiasoft/sirepo/sirepo/auth_db.py", line 217, in save
    self._session().commit()
  File "/home/vagrant/.pyenv/versions/3.7.2/envs/py3/lib/python3.7/site-packages/sqlalchemy/orm/session.py", line 1406, in commit
    self._transaction.commit(_to_root=self.future)
  File "/home/vagrant/.pyenv/versions/3.7.2/envs/py3/lib/python3.7/site-packages/sqlalchemy/orm/session.py", line 820, in commit
    self._prepare_impl()
  File "/home/vagrant/.pyenv/versions/3.7.2/envs/py3/lib/python3.7/site-packages/sqlalchemy/orm/session.py", line 799, in _prepare_impl
    self.session.flush()
  File "/home/vagrant/.pyenv/versions/3.7.2/envs/py3/lib/python3.7/site-packages/sqlalchemy/orm/session.py", line 3224, in flush
    self._flush(objects)
  File "/home/vagrant/.pyenv/versions/3.7.2/envs/py3/lib/python3.7/site-packages/sqlalchemy/orm/session.py", line 3364, in _flush
    transaction.rollback(_capture_exception=True)
  File "/home/vagrant/.pyenv/versions/3.7.2/envs/py3/lib/python3.7/site-packages/sqlalchemy/util/langhelpers.py", line 72, in __exit__
    with_traceback=exc_tb,
  File "/home/vagrant/.pyenv/versions/3.7.2/envs/py3/lib/python3.7/site-packages/sqlalchemy/util/compat.py", line 198, in raise_
    raise exception
  File "/home/vagrant/.pyenv/versions/3.7.2/envs/py3/lib/python3.7/site-packages/sqlalchemy/orm/session.py", line 3324, in _flush
    flush_context.execute()
  File "/home/vagrant/.pyenv/versions/3.7.2/envs/py3/lib/python3.7/site-packages/sqlalchemy/orm/unitofwork.py", line 453, in execute
    rec.execute(self)
  File "/home/vagrant/.pyenv/versions/3.7.2/envs/py3/lib/python3.7/site-packages/sqlalchemy/orm/unitofwork.py", line 630, in execute
    uow,
  File "/home/vagrant/.pyenv/versions/3.7.2/envs/py3/lib/python3.7/site-packages/sqlalchemy/orm/persistence.py", line 247, in save_obj
    insert,
  File "/home/vagrant/.pyenv/versions/3.7.2/envs/py3/lib/python3.7/site-packages/sqlalchemy/orm/persistence.py", line 1095, in _emit_insert_statements
    statement, multiparams, execution_options=execution_options
  File "/home/vagrant/.pyenv/versions/3.7.2/envs/py3/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1520, in _execute_20
    return meth(self, args_10style, kwargs_10style, execution_options)
  File "/home/vagrant/.pyenv/versions/3.7.2/envs/py3/lib/python3.7/site-packages/sqlalchemy/sql/elements.py", line 314, in _execute_on_connection
    self, multiparams, params, execution_options
  File "/home/vagrant/.pyenv/versions/3.7.2/envs/py3/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1399, in _execute_clauseelement
    cache_hit=cache_hit,
  File "/home/vagrant/.pyenv/versions/3.7.2/envs/py3/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1749, in _execute_context
    e, statement, parameters, cursor, context
  File "/home/vagrant/.pyenv/versions/3.7.2/envs/py3/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1930, in _handle_dbapi_exception
    sqlalchemy_exception, with_traceback=exc_info[2], from_=e
  File "/home/vagrant/.pyenv/versions/3.7.2/envs/py3/lib/python3.7/site-packages/sqlalchemy/util/compat.py", line 198, in raise_
    raise exception
  File "/home/vagrant/.pyenv/versions/3.7.2/envs/py3/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1706, in _execute_context
    cursor, statement, parameters, context
  File "/home/vagrant/.pyenv/versions/3.7.2/envs/py3/lib/python3.7/site-packages/sqlalchemy/engine/default.py", line 681, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.IntegrityError: (sqlite3.IntegrityError) UNIQUE constraint failed: user_role_invite_t.uid, user_role_invite_t.role
[SQL: INSERT INTO user_role_invite_t (uid, role, status, token, moderator_uid) VALUES (?, ?, ?, ?, ?)]
[parameters: ('cdnqkXeR', 'sim_type_jupyterhublogin', <ModerationStatus.PENDING: 'pending'>, 'ELzJHoSXZQdfft5FPLyjxEisDoUlBBrf', None)]
(Background on this error at: http://sqlalche.me/e/14/gkpj)

Exception was printed at:

sqlalchemy.exc.IntegrityError: (sqlite3.IntegrityError) UNIQUE constraint failed: user_role_invite_t.uid, user_role_invite_t.role
[SQL: INSERT INTO user_role_invite_t (uid, role, status, token, moderator_uid) VALUES (?, ?, ?, ?, ?)]
[parameters: ('cdnqkXeR', 'sim_type_jupyterhublogin', <ModerationStatus.PENDING: 'pending'>, 'ELzJHoSXZQdfft5FPLyjxEisDoUlBBrf', None)]
(Background on this error at: http://sqlalche.me/e/14/gkpj)
  File "/home/vagrant/.pyenv/versions/3.7.2/envs/py3/lib/python3.7/site-packages/flask/app.py", line 2091, in __call__
    return self.wsgi_app(environ, start_response)
  File "/home/vagrant/.pyenv/versions/3.7.2/envs/py3/lib/python3.7/site-packages/flask/app.py", line 2073, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/vagrant/.pyenv/versions/3.7.2/envs/py3/lib/python3.7/site-packages/flask/app.py", line 1516, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/vagrant/.pyenv/versions/3.7.2/envs/py3/lib/python3.7/site-packages/flask/app.py", line 1502, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
  File "/home/vagrant/src/radiasoft/sirepo/sirepo/uri_router.py", line 309, in _dispatch
    return _call_api(None, route, kwargs=kwargs)
  File "/home/vagrant/src/radiasoft/sirepo/sirepo/uri_router.py", line 269, in _call_api
    r = _response(getattr(qcall, qcall.uri_route.func_name)(**kwargs))
  File "/home/vagrant/src/radiasoft/sirepo/sirepo/quest.py", line 312, in _wrapper
    return self.func(*args, **kwargs)
'ip0 - li-uid2 "POST /save-moderation-reason HTTP/1.1" 200 233 "http://v.radia.run:8080/jupyterhublogin" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/ip1 Safari/537.36"'
```

**Comment by robnagler** [comment]
This will be "semi fixed" with PR #5197, because when the exception is thrown, the db will be rolled back and cascaded to the user (hopefully).
